Content.Shared/Turnstile/Systems/SharedTurnstileSystem.cs

```diff
@@ -0,0 +1,230 @@
+            Audio.PlayPvs(turnstile.BumpSound, ent.Owner, AudioParams.Default.WithVolume(-3));
+        }
+    }
+
+    /// <summary>
+    ///     Makes a turnstile proceed to the next state.
+    /// </summary>
+    private void NextState(Entity<TurnstileComponent> ent, TimeSpan time)
+    {
+        var turnstile = ent.Comp;
+
+        if (turnstile.State == TurnstileState.Rotating)
+        {
+            // Check and see if we're still colliding with the admitted entity.
+
+            var stillCollidingWithAdmitted = GetIsCollidingWithAdmitted(ent);
+            if (!stillCollidingWithAdmitted)
+            {
+                turnstile.NextStateChange = null;
+                SetState(ent, TurnstileState.Idle);
+                RemComp<PreventCollideComponent>(ent.Owner);
+            }
+            else
+            {
+                turnstile.NextStateChange = GameTiming.CurTime + turnstile.TurnstileTurnTime;
+            }
+        }
+    }
+
+    private void OnRemove(Entity<TurnstileComponent> turnstile, ref ComponentRemove args)
+    {
+        _activeTurnstiles.Remove(turnstile);
+    }
+
+    private bool GetIsCollidingWithAdmitted(Entity<TurnstileComponent, PreventCollideComponent?> ent)
+    {
+        if (!Resolve(ent.Owner, ref ent.Comp2))
+            return false;
+
+        var turnstileAABB = EntityLookupSystem.GetWorldAABB(ent.Owner);
+        var otherAABB = EntityLookupSystem.GetWorldAABB(ent.Comp2!.Uid);
+
+        if (turnstileAABB.Intersects(otherAABB))
+            return true;
+
+        return false;
+    }
+
+    /// <summary>
+    ///     Iterate over active turnstiles and progress them to the next state if they need to be updated.
+    /// </summary>
+    public override void Update(float frameTime)
+    {
+        var time = GameTiming.CurTime;
+        foreach (var ent in _activeTurnstiles.ToList())
+        {
+            SetCollidable(ent, true);
+            var turnstile = ent.Comp;
+            if (turnstile.Deleted || turnstile.NextStateChange == null)
+            {
+                _activeTurnstiles.Remove(ent);
+                continue;
+            }
+
+            if (Paused(ent))
+                continue;
+
+            if (turnstile.NextStateChange.Value < time)
+                NextState(ent, time);
+        }
+    }
+
+    protected Direction GetFacingDirection(Entity<TurnstileComponent> ent)
+    {
+        var xform = _xformQuery.GetComponent(ent.Owner);
+        return xform.LocalRotation.GetDir();
+    }
+
+    protected Direction GetDirectionOfContact(EntityUid uid, EntityUid other)
+    {
+        var xform = _xformQuery.GetComponent(uid);
+        var xformOther = _xformQuery.GetComponent(other);
+        return (xform.LocalPosition - xformOther.LocalPosition).GetDir();
+    }
+
+    private void SetCollidable(Entity<TurnstileComponent, PhysicsComponent?> ent, bool collidable)
+    {
+        if (!Resolve(ent.Owner, ref ent.Comp2))
```

Comment:
```suggestion
        if (!Resolve(ent, ref ent.Comp2))
```
`Entity` automatically converts into `EntityUid`. You usually don't need `.Owner` unless there is type ambiguity. (Lots of places, not just this line)